chain/chain/src/state_sync.rs

```diff
@@ -0,0 +1,191 @@
+        .map(|item| item.and_then(|(k, _v)| EpochId::try_from_slice(&k)))
+}
+
+fn save_epoch_new_chunks<T: ChainStoreAccess>(
+    chain_store: &T,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+) -> Result<(), Error> {
+    let Some(mut num_new_chunks) = get_epoch_new_chunks(chain_store.store(), header.prev_hash())?
+    else {
+        // This might happen in the case of epoch sync where we save individual headers without having all
+        // headers that belong to the epoch.
+        return Ok(());
+    };
+
+    // This shouldn't happen because block headers in the same epoch should have chunks masks
+    // of the same length, but we log it here in case it happens for some reason. We return Ok because if this
+    // happens, it's some bug in this state sync logic, because the chunk mask length of headers are checked when
+    // they're verified. So in this case we shouldn't fail to commit this store update and store this block header
+    if num_new_chunks.len() != header.chunk_mask().len() {
+        tracing::error!(
+            block_hash=%header.hash(), chunk_mask_len=%header.chunk_mask().len(), stored_len=%num_new_chunks.len(),
+            "block header's chunk mask not of the same length as stored value in DBCol::StateSyncNewChunks",
+        );
+        return Ok(());
+    }
+
+    let mut done = true;
+    for (shard_index, num_new_chunks) in num_new_chunks.iter_mut().enumerate() {
+        let new_chunk = header.chunk_mask()[shard_index];
+        if new_chunk {
+            *num_new_chunks += 1;
+        }
+        if *num_new_chunks < 2 {
+            done = false;
+        }
+    }
+    if done {
+        // TODO(current_epoch_state_sync): this will not be correct if this block doesn't end up finalized on the main chain.
+        // We should fix it by setting the sync hash when it's finalized, which requires making changes to how we take state snapshots.
+        store_update.set_ser(DBCol::StateSyncHashes, header.epoch_id().as_ref(), header.hash())?;
+        store_update.delete_all(DBCol::StateSyncNewChunks);
+        // TODO: remove old ones
+        return Ok(());
+    }
+
+    store_update.set_ser(DBCol::StateSyncNewChunks, header.hash().as_ref(), &num_new_chunks)?;
+    Ok(())
+}
+
+fn on_new_epoch(store_update: &mut StoreUpdate, header: &BlockHeader) -> Result<(), Error> {
+    let num_new_chunks = vec![0usize; header.chunk_mask().len()];
+    store_update.set_ser(DBCol::StateSyncNewChunks, header.hash().as_ref(), &num_new_chunks)?;
+    Ok(())
+}
+
+fn remove_old_epochs(
+    store: &Store,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+    prev_header: &BlockHeader,
+) -> Result<(), Error> {
+    for epoch_id in iter_state_sync_hashes_keys(store) {
+        let epoch_id = epoch_id?;
+        if &epoch_id != header.epoch_id() && &epoch_id != prev_header.epoch_id() {
+            store_update.delete(DBCol::StateSyncHashes, epoch_id.as_ref());
```

Comment:
What about the other column?

Reply:
I guess we could add logic to clean that one up here too, but we already remove the ones that have height lower than the final block's height, so there's currently only one place where we do that and I think it should be sufficient to keep that column small?